### PR TITLE
promql: extend test scripting language to support range queries

### PR DESCRIPTION
**Pull request by charleskorn**
This PR extends the PromQL test scripting language to support range queries in addition to the existing support for instant queries.

It also:
* adds tests for the language, including for instant queries
* improves the error messages returned in some cases (eg. when an invalid duration value is given as timestamp for an instant query, it'll now include more context about which eval command is problematic)

This was previously discussed on the CNCF Slack [here](https://cloud-native.slack.com/archives/C01AUBA4PFE/p1710480869547059).

### CODE REVIEW

File: promql/test.go [hunk 1]
```diff
@@ -578,78 +742,116 @@ func (t *test) exec(tc testCommand, engine engineQuerier) error {
...
+		if err != nil {
+			return err
+		}
+		rangeRes := q.Exec(t.context)
+		if rangeRes.Err != nil {
+			return fmt.Errorf("error evaluating query %q (line %d) in range mode: %w", iq.expr, cmd.line, rangeRes.Err)
+		}
+		defer q.Close()
+		if cmd.ordered {
+			// Ordering isn't defined for range queries.
```
Comment (bboreham) [inline promql/test.go:825]: Range queries are always ordered by labels - should it check that?

File: promql/test.go [hunk 2]
```diff
@@ -209,13 +232,50 @@ func (t *test) parseEval(lines []string, i int) (int, *evalCmd, error) {
...
+		}
+
+		cmd = newRangeEvalCmd(expr, testStartTime.Add(time.Duration(parsedFrom)), testStartTime.Add(time.Duration(parsedTo)), time.Duration(parsedStep), i+1)
 	}
-	ts := testStartTime.Add(time.Duration(offset))
 
-	cmd := newEvalCmd(expr, ts, i+1)
 	switch mod {
 	case "ordered":
 		cmd.ordered = true
```
Comment (bboreham) [inline promql/test.go:281]: Should it error here on a range query?

**Review by bboreham — APPROVED** [review]
LGTM; couple of thoughts.

I especially like that you added `TestRunTest`, which serves as documentation of a sort.

File: promql/test.go (hunk 2 above)
Comment (charleskorn) [reply promql/test.go:283]: Specifying `ordered` for a range query is prevented by the `patEvalRange` regex. I've added a comment here in 6a0dd50 to clarify.

File: promql/test.go (hunk 1 above)
Comment (charleskorn) [reply promql/test.go]: Makes sense to me, added in 977db40.

**Review by bwplotka — APPROVED** [review]
(Hello from the bug scrub meeting!)

LGTM, solid change to me, we can always improve/fix later here (tests)

**Comment by bwplotka** [comment]
Thank You!

**Pull request merged**